Improve organelle upgrade cost display and minor custom window fixes

Kasterisk opened the pull request:
**Brief Description of What This PR Does**

The cost for individual upgrades in the organelle upgrade GUI now takes into account the editor action history as an improvement to #4281. Due to this, it's perfectly possible for an upgrade to have a negative MP cost, this means that such action gives more MP instead and so must be conveyed correctly in the UI.

While working on that I came across a minor bug in the modal system which I also had to fix as it affects the organelle upgrade GUI. Further extensive testing is warranted to make sure no more problems come up with this.

**Related Issues**

Continuation of #4281.

<!-- List all issues this PR closes here with the closes syntax: 
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
If this is not related to an issue, it should be described in more detail why this PR is needed here instead.
-->

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

hhyyrylainen commented on locale/en.po:
```diff
@@ -4789,10 +4791,15 @@ msgstr "population:"
+#: ../src/microbe_stage/editor/MicrobePartSelection.cs:165
+#: ../src/microbe_stage/editor/tooltips/SelectionMenuToolTip.cs:343
+msgid "POSITIVE_SIGNED_NUMBER"
+msgstr "+{0}"
```
I actually don't agree with this being made into a translation, mainly because I've used a simple "workaround" in the prototypes already (prepend + when number > 0). This is also not symmetrical with negative numbers as there's no way to override those with a translation.

Kasterisk replied:
Okay. My solution before this was just to replace - with +, but then I thought about the localization strictness we had.

hhyyrylainen replied:
For working with pure numbers, I think it doesn't make sense to add this kind of asymmetry here. Like I said, I'd do something like `"+" + Math.Abs(number).ToString(CurrentCulture)`

Kasterisk replied:
Changed.

hhyyrylainen commented on src/gui_common/CustomPopupMenu.cs:
```diff
@@ -44,33 +42,26 @@ public void RemapDynamicChildren()
-
-    protected virtual void ResolveNodeReferences()
-    {
+        tween.TweenCallback(this, "hide");
```
Is this intentional that this now calls "hide" instead of `nameof(Hide)`?

Kasterisk replied:
Yes. I can't find the link to the Godot docs reference right now but I think it stated that you must use the native name when calling engine functions like this.

hhyyrylainen replied:
If that's the intention, that is correct, but reading the existing code, this used to send the Close method call to our code. Why the change from that to the godot inbuilt method?

Kasterisk replied:
Because we no longer have to set `closing` to false, `OnFolded` isn't needed, one statement only.

Kasterisk replied:
> Or is this an animation specific use case where the hide needs to be called?

Kind of, or any other thing you'd want the window to do when you call `Close` to it really.

https://github.com/Revolutionary-Games/Thrive/blob/4cb9670399c902dd7ebea2feb04bbbe543ef94a0/src/gui_common/CustomWindow.cs#L225-L238

> "OnHideAnimationComplete"

Already covered functionally by `OnHidden`.

hhyyrylainen replied:
Then why does this not call that `OnHidden`?

My point here is that a custom functionality we are adding to the windows, should use our own methods to control the program flow, not rely on calling a Godot inbuilt method which then through more complicated callbacks eventually does the right thing. That's my point that I'd much prefer to still call one of our own methods here rather than "hide".

Kasterisk replied:
Okay.

hhyyrylainen commented on src/gui_common/ModalManager.cs:
```diff
@@ -15,7 +16,9 @@ public class ModalManager : NodeWithInput
-    private readonly Stack<CustomWindow> modalStack = new();
+    private readonly Queue<CustomWindow> demotedModals = new();
+
+    private Stack<CustomWindow> modalStack = new();
```
Can this be readonly as well?

Kasterisk replied:
No, is used like this just few lines down
https://github.com/Revolutionary-Games/Thrive/blob/4cb9670399c902dd7ebea2feb04bbbe543ef94a0/src/gui_common/ModalManager.cs#L212-L221

hhyyrylainen replied:
Well if that is a common problem, I think maybe a Deque should be used then instead to allow the expected error condition to be handled correctly?

Kasterisk replied:
The situation can occur somewhere in the middle of the stack not just both ends. Not only that, we'd have to reimplement stack behavior

> Well if that is a common problem

I wouldn't say it's that common, and one line of LINQ for it shouldn't be much of a performance sap.

hhyyrylainen replied:
Conceptually if the thing doesn't act like a stack or queue, then those data structures shouldn't be used. A deque sounds exactly what is needed here (we already have a Deque library), it supports push and pop to both front and back, while also supporting access to removing middle elements.

hhyyrylainen commented on src/microbe_stage/editor/upgrades/OrganelleUpgradeGUI.cs:
```diff
@@ -113,7 +112,16 @@ public override void _Ready()
                 var selectionButton = upgradeSelectionButtonScene.Instance<MicrobePartSelection>();
 
-                var cost = (int)Math.Min(upgrade.MPCost * costMultiplier, 100);
+                var oldUpgrade = organelle.Upgrades ?? new OrganelleUpgrades();
```
This should be outside the loop to avoid unnecessary object creation.

hhyyrylainen commented:
I fixed the code style problems I saw, but now with playtesting I found out that https://github.com/Revolutionary-Games/Thrive/issues/4201 seems to consistently happen to me 100% of the time the first time I open the art gallery when I've restarted the game.

Also I noticed that very frequently the starting of a slideshow fails to fullscreen (I've never seen this problem before):
![2023-04-14_14 58 58 5072](https://user-images.githubusercontent.com/3796411/232038299-77b84a21-5b5d-40cf-8d82-89ec251b886c.png)

The slideshow plays only on the first image area and doesn't cover the entire screen.

hhyyrylainen commented:
Not sure if this was fixed but when trying to modify a cilia (on master branch) I got both the modify popup and the organelle popup open at the same time. I don't remember seeing that happen when quickly testing this, so I'd guess that it was fixed already?

Kasterisk commented:
> Not sure if this was fixed but when trying to modify a cilia (on master branch) I got both the modify popup and the organelle popup open at the same time.

That's what this PR was trying to fix :)

> Also I noticed that very frequently the starting of a slideshow fails to fullscreen (I've never seen this problem before):

That's new, I guess one of the recent commits broke it.

Kasterisk commented:
Alright, everything should be fixed. Tooltips for cilia upgrades should work, slidescreen and modify popup no longer stuck.

So, the cause for why the tooltips won't show was because back when I implemented the custom tooltip system, I've made it so that when tooltipable controls went out of the tree, their relationship with their associated tooltips is severed. This is of course conflicting with the new modal system, because modal windows exiting the tree is an integral part of how it works (reparenting), I fixed this by checking whether a tooltipable goes out of tree due to reparenting or other reasons.

Also made `ToolTipManager`'s canvas layer be 128 to allow tooltips to render above the modals and the screen overlays (because it comes after the screen overlays in the autoload hierarchy).

hhyyrylainen reviewed (approved):
This seems to now at least not bug all the time. I'll merge this so that I can now try to make the different approach for the reparenting.
